vllm/worker/model_runner.py

```diff
@@ -155,6 +155,238 @@ def from_broadcasted_tensor_dict(
+        for query_len in self.query_lens[-len(seq_ids):]:
+            if self.enable_lora:
+                self.lora_index_mapping += [lora_id] * query_len
+                self.lora_prompt_mapping.extend(
+                    [lora_id] *
+                    (query_len if seq_group_metadata.sampling_params
+                     and seq_group_metadata.sampling_params.prompt_logprobs
+                     is not None else 1))
+
+            mm_data = seq_group_metadata.multi_modal_data
+            if mm_data:
+                # Process multi-modal data
+                mm_kwargs = self.multi_modal_input_mapper(mm_data)
+                self.multi_modal_inputs_list.append(mm_kwargs)
+
+    def build(self, runner: "GPUModelRunnerBase") -> ModelInputForGPU:
+        if not self.input_tokens:
+            return self.model_input_cls()
+
+        batch_size = len(self.input_tokens)
+        use_captured_graph = (
+            self.decode_only and not runner.model_config.enforce_eager
+            and batch_size <= _BATCH_SIZES_TO_CAPTURE[-1]
+            and self.max_decode_seq_len <= runner.max_seq_len_to_capture)
+
+        # If cuda graph can be used, pad tensors accordingly.
+        # See `capture_model` API for more details.
+        # vLLM uses cuda graph only for decoding requests.
+        cuda_graph_pad_size = 0
+        if use_captured_graph:
+            graph_batch_size = _get_graph_batch_size(batch_size)
+            assert graph_batch_size >= batch_size
+            cuda_graph_pad_size = graph_batch_size - batch_size
+            batch_size = graph_batch_size
+
+        #### Tokens and positions.
+        self.input_tokens.extend([0] * cuda_graph_pad_size)
+        self.input_positions.extend([0] * cuda_graph_pad_size)
+        input_tokens_tensor = torch.tensor(self.input_tokens,
+                                           dtype=torch.long,
+                                           device=runner.device)
+        input_positions_tensor = torch.tensor(self.input_positions,
+                                              dtype=torch.long,
+                                              device=runner.device)
+
+        #### Sequence and query lengths.
+        if use_captured_graph:
+            self.seq_lens.extend([1] * cuda_graph_pad_size)
+
+        #### Attention metadata.
+        attn_metadata = self.attn_metadata_builder.build(
+            runner, self.seq_lens, self.query_lens, use_captured_graph,
+            cuda_graph_pad_size, batch_size)
+
+        #### LoRA and multi-modal data.
+        if self.enable_lora:
+            self.lora_index_mapping.extend([0] * cuda_graph_pad_size)
+            lora_mapping = LoRAMapping(
+                self.lora_index_mapping,
+                self.lora_prompt_mapping,
+            )
+        else:
+            lora_mapping = None
+
+        multi_modal_kwargs = MultiModalInputs.batch(
+            self.multi_modal_inputs_list, device=runner.device)
+
+        return self.model_input_cls(
```

Comment:
can we have a `to(device)` API? before this call, all the prepared input lives in cpu.

Reply:
So you suggest like the following?

```python
        builder = ModelInputForGPUBuilder(...)
        for seq_group_metadata in seq_group_metadata_list:
            builder.add_seq_group(seq_group_metadata)
        model_input = builder.build(self) # All tensors are on CPU.
        return model_input.to(device) # Move to GPU.
```
This would introduce some overheads as we create tensors on CPU and then move to GPU, instead of directly creating them on GPU.

Reply:
yes, this would simplify the control plane communication.

actually, if you directly creating them on GPU, pytorch still needs to prepare them in cpu, and then move it to GPU together. I don't think there would be significant overhead here.

Reply:
The following code snippets are essentially the same:

```python
data = list(range(1000))
gpu_tensor = torch.tensor(data, dtype=torch.int, device="cuda")
```

```python
data = list(range(1000))
cpu_tensor = torch.tensor(data, dtype=torch.int, device="cpu")
gpu_tensor = cpu_tensor.to("cuda")
```

In the first code, pytorch still needs to create a cpu array to hold the Python data structure, and then move it to GPU.

If we do the cpu-gpu movement before we broadcast it to the rest workers, they will wait for the broadcast.

timeline:
```
driver worker: |  prepare input in cpu | move to GPU | broadcast |
rest workers:  |    wait for broadcast                         |
```

If we do the cpu-gpu movement after we broadcast, the driver worker sends cpu data (which can be fast because we have shared memory transport), and all workers move data to gpu respectively.

timeline:
```
driver worker: |  prepare input in cpu | broadcast | move to GPU |
rest workers:  |    wait for broadcast           | move to GPU |
```
Because these data are usually quite small, I suppose the second approach would be faster. Broadcasting small cpu data would be easier. Broadcasting small GPU data with NCCL is not good in performance.

Reply:
Hmm...make sense since we will have a Python list anyways. I'll think about that and add this API if no other concerns. Thanks.

Reply:
I've basically done the first round of refactoring and feel this is doable, but I'd prefer to defer this until #6052 is merged, because I plan to add `.to()` to the following data classes:
1. ModelInputForGPU, which will implicitly call `attn_metadata.to(...)` (the next point)
2. AttentionMetadata <- conflict